File: 0024 Swap Nodes in Pairs/solution.cpp
#include <bits/stdc++.h>

using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        if (!head || !head -> next)
            return head;
        ListNode *dummy = new ListNode(0);
        dummy -> next = head;
        ListNode *p = dummy;
        
        while (p)
        {
            ListNode *first = p -> next;
            if (!first)  break;
            
            ListNode *second = first -> next;
            if (!second) break;
            
            p -> next = second;
            first -> next = second -> next;
            second -> next = first;
            
            p = first;
        }
        return dummy -> next;
    }
};

ListNode *create_linkedlist(std::initializer_list<int> lst)
{
    auto iter = lst.begin();
    ListNode *head = lst.size() ? new ListNode(*iter++) : nullptr;
    for (ListNode *cur=head; iter != lst.end(); cur=cur->next)
        cur->next = new ListNode(*iter++);
    return head;
}

void clear(ListNode *head)
{
    while (head) {
        ListNode *del = head;
        head = head->next;
        delete del;
    }
}

void printList(ListNode *head) {
    for (ListNode *cur = head; cur; cur = cur->next)
        cout << cur->val << "->";
    cout << "\b\b  " << endl;
}

int main()
{
    Solution s;
    ListNode *head = create_linkedlist({1,2,3,4});
    ListNode *ret = s.swapPairs(head);
    printList(ret);
    clear(ret);

    return 0;
}

File: 0024 Swap Nodes in Pairs/solution.py
#! python3

class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def swapPairs(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        dummy = p1 = ListNode(0)
        dummy.next = head
        
        try:
            while True:
                p0, p1, p2 = p1, p1.next, p1.next.next
                p0.next, p1.next, p2.next = p2, p2.next, p1
        except:
            return dummy.next

def createLinkedlist(lst):
    l = len(lst)
    if not l:
        return None

    head = ListNode(lst[0])
    cur = head
    for i in range(1, l):
        cur.next = ListNode(lst[i])
        i += 1
        cur = cur.next
    return head

def printList(head):
    while head:
        print(head.val, end='->')
        head = head.next
    print("\b\b  ")

if __name__ == "__main__":
    head = createLinkedlist([1,2,3,4])
    res = Solution().swapPairs(head)
    printList(res)

